Adds a "Download CSV" button to the Learner List View

pomegranited opened the pull request:
Adds a button to the Course Learners page, which allows the selected learners list to be downloaded in CSV format.

This report is proxied as a plain text response from the Learner Analytics Data API.

This feature is disabled by default, and can be enabled using the `enable_learner_download` waffle switch.

**Discussions**: Architecture discussed extensively with @mulby and the analytics team, and UI changes designed by Alyssa Boehm.

**Dependencies**: [edx-analytics-data-api #128](https://github.com/edx/edx-analytics-data-api/pull/128)

**Screenshots**:

With no search filters:

![screen shot 2016-08-10 at 5 13 12 pm](https://cloud.githubusercontent.com/assets/7556571/17545776/cc822a2c-5f1d-11e6-9e40-8d82675e5c7a.png)

With search filters:
![screen shot 2016-08-10 at 5 14 15 pm](https://cloud.githubusercontent.com/assets/7556571/17545797/f1ad83a0-5f1d-11e6-8bff-fac0bc3e8657.png)

Responsive positioning on small screen:
![screen shot 2016-08-10 at 5 15 47 pm](https://cloud.githubusercontent.com/assets/7556571/17545831/1e1dcaa8-5f1e-11e6-9813-5b37c7d1cc75.png)

**Partner information**: 3rd party-hosted open edX instance

**Testing instructions**:

1. See testing setup instructions on [edx-analytics-data-api #128](https://github.com/edx/edx-analytics-data-api/pull/128).
1. Run the LMS, Analytics elasticsearch service, and analytics_data_api.  Ensure that you can see your course in the Analytics API http://localhost:8100/docs/#!/api/Learner_List.
1. Update `/edx/etc/insights.xml` to store your Analytics API client token, and your desired CSV field list, e.g.,

    ```
DATA_API_AUTH_TOKEN: your-api-token
````
1. Enable the Learner view in Insights:

    ```
sudo su insights
./manage.py waffle_flag display_learner_analytics --staff --create
./manage.py waffle_switch enable_course_api on --create
```
1. Launch Insights, visit the Learners page, and ensure that there is no `Download CSV` button like the one in the above screenshots.
1. Enable this feature switch:

    ```
./manage.py waffle_switch enable_learner_download on --create 
```
1. Relaunch Insights and visit the Learners page to see the new `Download CSV` button.  Click on it to download the CSV file, which should contain all the fields in the learner API call, in alphabetical order.
1. Update `/edx/etc/insights.xml` to update the CSV field list, e.g.,

    ```
LEARNER_API_LIST_DOWNLOAD_FIELDS: username,name,email,cohort,enrollment_date,last_updated
```
1. Relaunch Insights, and download the CSV report again.  It should contain only the fields specified in your `settings.LEARNER_API_LIST_DOWNLOAD_FIELDS`, in that order.
1. Load the CSV report URL in an incognito browser, and ensure that you see a Permission Denied error.

**Reviewers**
- [x] @bradenmacdonald 
- [x] @dsjen 
- [ ] @thallada 
- [x] Accessibility - @clrux

**Pre-Merge Checklist**

- [x] Rebase with current master
- [x] Squash into single commit
- [ ] Ensure tests pass
- [ ] Merge

edx-webhook commented:
Thanks for the pull request, @pomegranited! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

To automatically create an OSPR issue for this pull request, just visit this link: https://openedx-webhooks.herokuapp.com/github/process_pr?number=533&repo=edx%2Fedx-analytics-dashboard

edx-webhook commented:
Thanks for the pull request, @pomegranited! I've created [OSPR-1389](https://openedx.atlassian.net/browse/OSPR-1389) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).



If you like, you can add yourself to the [AUTHORS](https://github.com/edx/edx-analytics-dashboard/blob/master/AUTHORS)      file for this repo, though that isn't required.     Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

gsong commented:
@pomegranited Is this still WIP? If not, can you remove the designation from the title and we'll jump on reviewing it? Thanks.

nedbat commented:
@pomegranited btw, @gsong is another new dev advocate here at edX, and this was the first PR he commented on! :)

pomegranited commented:
Welcome @gsong!  Looking forward to working with you.

This PR is still WIP, but will be done and through our internal review by Monday AM.  I'll be sure to fix up the title and remove the 'waiting on author' flag from the OSPR when it's ready.

Thank you!

bradenmacdonald commented on analytics_dashboard/courses/views/learners.py:
```diff
@@ -59,4 +63,15 @@ def get_context_data(self, **kwargs):
+            list_fields = getattr(settings, 'LEARNER_API_LIST_DOWNLOAD_FIELDS', None)
+            if list_fields is not None:
+                list_download_url = '{}?{}'.format(list_download_url,
+                                                   urlencode(dict(fields=list_fields), doseq=True))
```
I don't think you want `doseq=True` here?

1. `list_fields` is a string, so it has no effect.
2. If `list_fields` were a list like `['username', 'name']` (which I don't think it is), this would result in `'fields=username&fields=name'`, which I don't think is what you want.

pomegranited replied:
Oo, good point!  Addressed with [5caf783](https://github.com/edx/edx-analytics-dashboard/pull/533/commits/5caf783c619f8831b12bcb81637e4ce5178dbf48).

bradenmacdonald commented on analytics_dashboard/static/apps/learners/app/app.js:
```diff
@@ -28,6 +28,8 @@ define(function(require) {
          *   for the HTML element that this app should attach to
          * - learnerListUrl (string) required - the URL for the
          *   Learner List API endpoint.
+         * - learnerListDownloadUrl (string) required - the download URL
```
"the CSV download URL" perhaps?

pomegranited replied:
Addressed with [5caf783](https://github.com/edx/edx-analytics-dashboard/pull/533/commits/5caf783c619f8831b12bcb81637e4ce5178dbf48).

bradenmacdonald commented on analytics_dashboard/static/apps/learners/common/templates/download-data.underscore:
```diff
@@ -0,0 +1,10 @@
+    <div class="section-action">
+        <a class="btn btn-default action-download-data"
+           data-role="learner-data-csv" data-track-type="click"
+           data-track-event="edx.bi.csv.downloaded" data-track-category="trend">
```
I believe `data-track-category="trend"` is used for the content engagement report download, and this should be some new value like `data-track-category="learner-profiles"`. Not sure though.

bradenmacdonald replied:
For some reason, I can't tab to this button with the keyboard. I think there's some `tabindex` issue or similar.

pomegranited replied:
I believe I've fixed this with [5caf783](https://github.com/edx/edx-analytics-dashboard/pull/533/commits/5caf783c619f8831b12bcb81637e4ce5178dbf48), by setting a default href="#" on the link.

Also made the `data-track-category` configurable.

bradenmacdonald commented on analytics_dashboard/static/apps/learners/common/views/download-data.js:
```diff
@@ -0,0 +1,73 @@
+            // Parse the download_url, if given, into URL and query params
+            if (this.options.collection && this.options.collection.download_url) {
+                splitUrl = this.options.collection.download_url.split('?', 2);
+                this.download_url = splitUrl[0];
```
I think it's potentially confusing that `DownloadDataView.download_url` and `DownloadDataView.getDownloadUrl()` both seem like public properties but return different things. 

Normally I would suggest that `download_url` and `download_param` should be prefixed with `_` or marked as private via some mechanism like that, but it seems like this codebase doesn't use those conventions. So perhaps just rename `this.download_url` to `this.download_base_url` ?

pomegranited replied:
Addressed with [5caf783](https://github.com/edx/edx-analytics-dashboard/pull/533/commits/5caf783c619f8831b12bcb81637e4ce5178dbf48).

bradenmacdonald commented on analytics_dashboard/static/apps/learners/common/views/download-data.js:
```diff
@@ -0,0 +1,73 @@
+                orderedParams = [];
+
+            // Return empty string if no download_url
+            if (_.isEmpty(this.download_url)) {
```
Not an issue, but a question: Earlier you just used `if (this.options.collection.download_url)` to set `this.download_url` so here is this any different than just using `if (!this.download_url) {` ?

pomegranited replied:
Using `_.isEmpty` returns true for for 0, null and empty string, as well as empty lists and objects, so I should use that instead here, since [] and {} aren't valid URLs, though it's unlikely we'd receive one of these.

Also, I've been trying to be more conscious of not assuming language-specific boolean values, so will change the check above to use `_.isEmpty`.

bradenmacdonald commented on analytics_dashboard/static/apps/learners/common/views/download-data.js:
```diff
@@ -0,0 +1,73 @@
+                    key: key,
+                    val: queryParams[key]
+                });
+            });
```
Very optional nit: If you want, you can do this with native ES5:

```
var orderedParams = Object.keys(queryParams).sort().map(function (key) {
     return {key: key, val: queryParams[key]}
});
```

And because I love ES6 I can't help but point out that soon we'll be able to do
```
const orderedParams = Object.keys(queryParams).sort().map(key => ({key, val: queryParams[key]}));
```

pomegranited replied:
I do want to do that!  I'm not familiar enough with ES5, and spent entirely too long trying to figure out how to minimize that bit of code.  Yours is much prettier! 

ES6 will be excellent too..

bradenmacdonald commented on analytics_dashboard/static/apps/learners/common/views/download-data.js:
```diff
@@ -0,0 +1,73 @@
+
+        downloadData: function(event) {
+            // Set the href for the download button when clicked
+            $(event.currentTarget).attr('href', this.getDownloadUrl());
```
This pattern of setting `href` dynamically isn't the best. If I middle-click or press ENTER on this button, nothing happens the first time. But if I left-click it and download a CSV, then middle-clicking or pressing ENTER does have the desired effect in the future (but won't match the latest filter selections).

pomegranited replied:
I've set a default href="#", and am updating the href when the collection changes.
See [5caf783](https://github.com/edx/edx-analytics-dashboard/pull/533/commits/5caf783c619f8831b12bcb81637e4ce5178dbf48).

pomegranited commented:
@bradenmacdonald Should be ready for another round of review?  *fingers crossed on CI*

bradenmacdonald commented on analytics_dashboard/static/apps/learners/common/views/download-data.js:
```diff
@@ -10,61 +10,63 @@ define(function(require) {
             }
+
+            // Update the href on the download button when the collection changes.
+            this.listenTo(this.options.collection, 'sync', this.updateDownloadLink);
```
This is nice, but there is a problem: this doesn't seem to be fired when the page loads. As a result, the button/link doesn't work unless the user first changes some filters.

bradenmacdonald commented on analytics_dashboard/static/apps/learners/common/views/download-data.js:
```diff
@@ -10,61 +10,63 @@ define(function(require) {
             var splitUrl;
-            this.options = options || {};
+            this.options = _.extend({
+                downloadLinkId: 'download-data',
```
From [these Marionette docs](http://marionettejs.com/docs/v2.4.7/marionette.itemview.html), it looks to me like a better way to do this (without an ID) could be to declare
```javascript
DownloadDataView = Marionette.ItemView.extend({
...
 ui: {
    csvDownloadButton: '.action-download-data',
  },
...
```

And then you can just manipulate `this.ui.csvDownloadButton`

e.g.
```
updateDownloadLink: function() {
    this.ui.csvDownloadButton.attr('href', this.getDownloadUrl());
}
```

bradenmacdonald commented:
Found one bug. As for the CI, it looks like you've just got some eslint errors which should be nice and easy to fix :)

pomegranited commented:
@bradenmacdonald Ach, thank you for finding that bug, and for showing me that I didn't need to set an ID on the link.  Tests are passing again, but even better, the feature is working for me now.

Does it look ok to you?

pomegranited commented:
Rebased [c88c065](https://github.com/open-craft/edx-analytics-dashboard/commit/c88c065a11bf27a6ec256941da0ddd394a1743b7) from [master](https://github.com/edx/edx-analytics-dashboard/commit/b85a61bc87b8aabc8aa8228a538a5e9276899e4b) and squashes commits into [8bcf729](https://github.com/open-craft/edx-analytics-dashboard/commit/8bcf72912060aa29bbc98aaf19c3003d2a1ed108).

bradenmacdonald commented:
:+1:

- [x] I tested this: I followed the testing instructions
- [x] I read through the code
- [x] Includes documentation: code contains docstrings; feature is intuitive and matches existing CSV download pattern used in other Insights reports

mulby commented:
@lamagnifica FYI - note that this is behind a feature flag so we can control when this becomes user visible.

lamagnifica commented:
@mulby thank you. Please tag me when you open an AN story to change the feature flag for edX Insights.

dsjen commented:
@pomegranited -- thanks for the PR!  We'll be reviewing this soon.

dsjen commented on analytics_dashboard/static/js/utils/utils.js:
```diff
@@ -138,6 +138,28 @@ define(['moment', 'underscore', 'utils/globalization'], function(moment, _, Glob
+            }).join('&');
+
+            // Prefix query string params with '?', but return an empty string if there are no params.
+            return fragment !== '' ? ('?' + fragment) : fragment;
```
Nit: I think this might read a bit better in the positive (e.g. `return fragment === '' ? fragment : ('?' + fragment);`) instead.

pomegranited replied:
Fixed with [beea697](https://github.com/open-craft/edx-analytics-dashboard/commit/beea697f557b3f22bded878835220c45c994fca3).

dsjen commented on analytics_dashboard/static/apps/learners/common/templates/download-data.underscore:
```diff
@@ -0,0 +1,10 @@
+<% if (hasDownloadData) { %>
+    <div class="section-action">
+        <a class="btn btn-default action-download-data"
```
You've caught us as we're actively updating our buttons to actually be buttons instead of links.  Please see https://github.com/edx/edx-analytics-dashboard/pull/539/files#diff-8d0d7afbb3f6a88852a537eb707307bcR11 for an example this work motivated by accessibility needs.

pomegranited replied:
@dsjen That's a shame, because as a link, you can right-click it and choose to save the file to a different location.

Is there another, more accessible option that lets us keep this link?  Same question applies to [PR #538](https://github.com/edx/edx-analytics-dashboard/pull/538/files#diff-da3307d62affb314f23f7431ecb22b50R22).

dsjen replied:
@pomegranited -- makes sense!  @edx/edx-accessibility, is there a way to accomplish both cases of making this semantically a button, but also enable the right-click save as?

clrux replied:
So, clicking the button/link would download? If so, why would we need a separate right-click to save as?

dsjen replied:
I believe @pomegranited 's use case is to save the file to a different location.

clrux replied:
If we use a button and set the MIME type for the target, it should prompt a normal download and let folks choose where they want to save the file. Do we have that option?

pomegranited replied:
@clrux Yep, the target sets the `Content-Type` for the returned response, and a `content-disposition` header which designates the link response as an attachment with a filename.

However, the "normal download" behaviour for a `<form>` containing a `<button>` is to download the file to the default download directory, not to prompt for a file save.  If it's a link, I can right-click on it, and save the file wherever I like.  Also, if it's a link, I can see the download URL before I click on it.

Also relevant (though not technically prohibitive) is that this "Download CSV" URL is constructed from concatenating a bunch of query string parameters.  To make this work as a `<form>` and `<button>`, I need to parse out these query string parameters, and dynamically create a `<input type="hidden">` element for each of them.  I can do this, but we still have a dynamic content-changing accessibility situation.

The other alternative is to have the `<button>` trigger a Javascript event that opens a new window with the download URL, but that still doesn't allow right-click-save-as or any of the other ways of clicking a link.

Can I please keep the link, styled as a button?

pomegranited replied:
@clrux I'm also happy to make my link look like a link instead of a button, if @edx/edx-ux-team concur.

clrux replied:
I'm wondering if we're confusing form/button behavior with browser default behavior. For example, I have to change my Chrome (or Firefox, Internet Explorer, Safari, etc.) settings to "Ask me where to save files" as it defaults to the "Downloads" folder without asking.

Since not everyone changes their settings and may wish to download to a specific location, I'm OK with the following:

* keep it a link, but give it a role of button (`role="button"`)
* style it to look like a button so it's consistent with other buttons

pomegranited replied:
Thank you , @clrux , for coming up with an accessible compromise!  I've committed the link, with `role="button"`, keeping the styling like other buttons.

See [3f86edd1 download-data.underscore L3](https://github.com/edx/edx-analytics-dashboard/pull/533/commits/3f86edd1a33c288cecd82799298b271b5c77c321#diff-369fecdc861ecb90eac428b16d0d8d70R3)

dsjen commented on analytics_dashboard/static/apps/learners/common/collections/learners.js:
```diff
@@ -16,6 +16,7 @@ define(function(require) {
             PagingCollection.prototype.initialize.call(this, options);
 
             this.url = options.url;
+            this.download_url = options.download_url;
```
Our JS convention is to use camelCase. :)

pomegranited replied:
Gah..  have to break my fingers' python habits to write JS 😄 

Fixed with [00efd18](https://github.com/open-craft/edx-analytics-dashboard/commit/00efd180a3184a3ce4edc78aa20eb80059616077).

dsjen commented on analytics_dashboard/static/apps/learners/app/app.js:
```diff
@@ -54,6 +56,7 @@ define(function(require) {
 
             learnerCollection = new LearnerCollection(this.options.learnerListJson, {
                 url: this.options.learnerListUrl,
+                download_url: this.options.learnerListDownloadUrl,
```
Our JS convention is to use camelCase. :)

pomegranited replied:
Fixed with [00efd18](https://github.com/open-craft/edx-analytics-dashboard/commit/00efd180a3184a3ce4edc78aa20eb80059616077).

dsjen commented on analytics_dashboard/courses/views/learners.py:
```diff
@@ -59,4 +63,15 @@ def get_context_data(self, **kwargs):
+            list_download_url = reverse('learner_analytics_api:v0:LearnerListCsv')
+
+            # Append the 'fields' parameter if configured
+            list_fields = getattr(settings, 'LEARNER_API_LIST_DOWNLOAD_FIELDS', None)
```
Can `LEARNER_API_LIST_DOWNLOAD_FIELDS` be added to [settings/base.py](https://github.com/edx/edx-analytics-dashboard/blob/master/analytics_dashboard/settings/base.py) too?  We typically add settings there even if set to `None` so that we know what is configurable.

pomegranited replied:
Fixed with [f02f95e](https://github.com/open-craft/edx-analytics-dashboard/commit/f02f95e8e4f770710086e63d4a4afb98d7124ca7).

dsjen commented on analytics_dashboard/courses/views/learners.py:
```diff
@@ -59,4 +63,15 @@ def get_context_data(self, **kwargs):
         context['show_error'] = False if context['learner_list_json'] else True
 
+        # Only show learner download button(s) if switch is enabled
+        if switch_is_active('enable_learner_download'):
```
Nice!  Please add documentation for this switch in our [readme](https://github.com/edx/edx-analytics-dashboard/blob/master/README.md#feature-gating) too!

pomegranited replied:
Fixed with [b745c0c](https://github.com/open-craft/edx-analytics-dashboard/commit/b745c0c612d6cbd9122d385fc003f6d232855257).

dsjen commented on analytics_dashboard/learner_analytics_api/v0/tests/test_views.py:
```diff
@@ -98,6 +105,30 @@ def test_no_course_id_provided(self):
 
 
+@ddt.ddt
+class LearnerListCsvViewTestCase(LearnerListViewTestCase):
```
Nit: We've been capitalizing "CSV", e.g. https://github.com/edx/edx-analytics-dashboard/blob/master/analytics_dashboard/courses/views/csv.py#L42. :)

pomegranited replied:
Fixed with [16864a4](https://github.com/open-craft/edx-analytics-dashboard/commit/16864a4c812ca9dc26b9342311972dc41f5baa74).

dsjen commented on analytics_dashboard/learner_analytics_api/v0/urls.py:
```diff
@@ -10,6 +10,7 @@
 urlpatterns = [
     url(r'^learners/{}/$'.format(USERNAME_PATTERN), views.LearnerDetailView.as_view(), name='LearnerDetail'),
     url(r'^learners/$', views.LearnerListView.as_view(), name='LearnerList'),
+    url(r'^learners.csv$', views.LearnerListCsvView.as_view(), name='LearnerListCsv'),
```
Nit: We've been capitalizing "CSV", e.g. https://github.com/edx/edx-analytics-dashboard/blob/master/analytics_dashboard/courses/views/csv.py#L42. :)

pomegranited replied:
Fixed with [16864a4](https://github.com/open-craft/edx-analytics-dashboard/commit/16864a4c812ca9dc26b9342311972dc41f5baa74).

dsjen commented on analytics_dashboard/learner_analytics_api/v0/views.py:
```diff
@@ -93,16 +133,24 @@ def not_found_developer_message(self):
+        return self.client.learners.get(**request.query_params)
+
+
+class LearnerListCsvView(DownloadLearnerApiViewMixin, LearnerListView):
```
Nit: We've been capitalizing "CSV", e.g. https://github.com/edx/edx-analytics-dashboard/blob/master/analytics_dashboard/courses/views/csv.py#L42. :)

pomegranited replied:
Fixed with [16864a4](https://github.com/open-craft/edx-analytics-dashboard/commit/16864a4c812ca9dc26b9342311972dc41f5baa74).

dsjen commented on analytics_dashboard/static/apps/learners/common/views/download-data.js:
```diff
@@ -0,0 +1,81 @@
+            }, options);
+
+            // Parse the download_url, if given, into URL and query params
+            if (this.options.collection && !_.isEmpty(this.options.collection.download_url)) {
```
Our convention is to use camelCase for JS. :)

pomegranited replied:
Fixed with [00efd18](https://github.com/open-craft/edx-analytics-dashboard/commit/00efd180a3184a3ce4edc78aa20eb80059616077).

dsjen commented on analytics_dashboard/static/apps/learners/common/views/download-data.js:
```diff
@@ -0,0 +1,81 @@
+
+        initialize: function(options) {
+            var splitUrl;
+            this.options = _.extend({
```
We probably want to make a copy of the object, similar to what is done in https://github.com/edx/edx-analytics-dashboard/blob/31338f537feb8cabbedae551550f3dc1f0d763b5/analytics_dashboard/static/apps/learners/common/views/alert-view.js#L59.

pomegranited replied:
Fixed with [9945f19](https://github.com/open-craft/edx-analytics-dashboard/commit/9945f193f0d206d7c6783c3e7e1429cf30e1cefa).

dsjen commented on analytics_dashboard/static/apps/learners/common/views/spec/download-data-spec.js:
```diff
@@ -0,0 +1,147 @@
+        });
+
+        it('has default options', function() {
+            var downloadDataView = new DownloadDataView({});
```
Our JS convention is to only use one `var` per function. :)

pomegranited replied:
Fixed with [6f84645](https://github.com/open-craft/edx-analytics-dashboard/commit/6f84645f0b88efd20d3af47ea062daaeca97c7cf).

dsjen commented on analytics_dashboard/static/apps/learners/common/views/spec/download-data-spec.js:
```diff
@@ -0,0 +1,147 @@
+                downloadDataTitle: 'Download My CSV',
+                downloadDataMessage: 'Download my search results'
+            });
+            var templateVars = downloadDataView.templateHelpers();
```
Our JS convention is to only use one `var` per function. :)

pomegranited replied:
Fixed with [6f84645](https://github.com/open-craft/edx-analytics-dashboard/commit/6f84645f0b88efd20d3af47ea062daaeca97c7cf).

dsjen commented on analytics_dashboard/static/apps/learners/common/views/spec/download-data-spec.js:
```diff
@@ -0,0 +1,147 @@
+                    {url: 'http://example.com'}
+                )
+            });
+            var templateVars = downloadDataView.templateHelpers();
```
Our JS convention is to only use one `var` per function. :)

pomegranited replied:
Fixed with [6f84645](https://github.com/open-craft/edx-analytics-dashboard/commit/6f84645f0b88efd20d3af47ea062daaeca97c7cf).

dsjen commented on analytics_dashboard/static/apps/learners/common/views/spec/download-data-spec.js:
```diff
@@ -0,0 +1,147 @@
+                    courseId: 'course-v1:Demo:test'
+                }
+            );
+            var downloadDataView = new DownloadDataView({
```
Our JS convention is to only use one `var` per function. :)

pomegranited replied:
Fixed with [6f84645](https://github.com/open-craft/edx-analytics-dashboard/commit/6f84645f0b88efd20d3af47ea062daaeca97c7cf).

dsjen commented on analytics_dashboard/learner_analytics_api/v0/views.py:
```diff
@@ -60,6 +77,29 @@ def handle_exception(self, exc):
+class DownloadLearnerApiViewMixin(object):
+    """
+    Requests text/csv data from the Learner Analytics API,
+     and ensures that the REST framework returns it unparsed,
```
Nit: there seem to be extra spaces on these lines.

dsjen replied:
Also, thank you for all the great comments!

pomegranited replied:
Glad you like them :smile:

Fixed spacing with [d7fb3c0](https://github.com/open-craft/edx-analytics-dashboard/commit/d7fb3c04356ead8f5f42dc92747055b83b109979).

dsjen commented on analytics_dashboard/static/apps/learners/common/views/spec/download-data-spec.js:
```diff
@@ -0,0 +1,147 @@
+    describe('DownloadDataView', function() {
+        beforeEach(function() {
+            this.user = {
+                last_updated: new Date(2016, 1, 28)
```
Our JS convention is to use camelCase. :)

pomegranited replied:
Didn't need `last_updated` at all, so just removed it with [00efd18](https://github.com/open-craft/edx-analytics-dashboard/commit/00efd180a3184a3ce4edc78aa20eb80059616077).

dsjen commented on analytics_dashboard/static/apps/learners/roster/views/roster.js:
```diff
@@ -62,6 +64,10 @@ define(function(require) {
             this.showChildView('activityDateRange', new ActiveDateRangeView({
                 model: this.options.courseMetadata
             }));
+            this.showChildView('downloadData', new DownloadDataView({
```
Could this be skipped if the download url isn't available?

pomegranited replied:
I'd rather let the DownloadDataView [decide when to show itself](https://github.com/edx/edx-analytics-dashboard/pull/533/files#diff-369fecdc861ecb90eac428b16d0d8d70R1), than make the RosterView have to be aware of the conditions that DownloadDataView shows up.

But I can be convinced otherwise :smile:.

dsjen replied:
Ah!  That makes sense.  👍

dsjen commented on analytics_dashboard/static/apps/learners/roster/views/roster.js:
```diff
@@ -62,6 +64,10 @@ define(function(require) {
             }));
+            this.showChildView('downloadData', new DownloadDataView({
+                collection: this.options.collection,
+                trackCategory: 'learner-profile'
```
@stroilova is this the preferred tracking category for when the "Download CSV" button is clicked when download the learner roster?

It looks like all together, the event is:

**edx.bi.csv.downloaded**
* type: "click"
* category: "learner-profile"

It looks like we've been using underscores with the categories too, according to https://github.com/edx/edx-analytics-dashboard/pull/542.

I would suggest something more like: `learners_list`, although I'd expect @stroilova to have a more thought out answer. 😄 

CC: @thallada (in case you have opinions about the events having just implemented a bunch of new ones).

thallada replied:
I think to be consistent with the changes in #542, the category here should be "learner_roster".

dsjen replied:
@pomegranited -- please add a test for the triggered event.  We do this in a few places, for example: https://github.com/edx/edx-analytics-dashboard/blob/master/analytics_dashboard/static/apps/learners/roster/views/spec/roster-spec.js#L457

pomegranited replied:
Changed the category to `learner_roster` with [2d44cba](https://github.com/open-craft/edx-analytics-dashboard/commit/2d44cba0fc1b672df294805c3c1fcb0225f6f6fc).

pomegranited replied:
@thallada @dsjen CC @stroilova 

Glad you asked me to unit test that event trigger, because it wasn't actually triggering!  I thought it was enough to have the [`data-track-*` attributes in place for in the template](https://github.com/open-craft/edx-analytics-dashboard/blob/a740d28265bd6fa3143d4e5afee4c8163a32e307/analytics_dashboard/static/apps/learners/common/templates/download-data.underscore#L3-L5), but as @dsjen pointed out, the link isn't present in the DOM when the page is first loaded, so none of the lovely automatic tracking is done for me.

Submitted [a740d28](https://github.com/edx/edx-analytics-dashboard/pull/533/commits/a740d28265bd6fa3143d4e5afee4c8163a32e307) which triggers the event explicitly on button click, and adds unit tests to make sure.

dsjen replied:
Wonderful!!!  Thank you for the updates and for dealing with our dynamic DOM.  🎆

dsjen commented:
@pomegranited -- I'm trying to get this working, but am running into an error (`AssertionError: Hop-by-hop headers not allowed`):

```
2016-08-23 18:08:54,380 DEBUG 94449 [requests.packages.urllib3.connectionpool] connectionpool.py:395 - "GET /api/v0/learners/?course_id=course-v1%3AHarvardX%2BSW25x.1%2B2015&enrollment_mode=honor&ignore_segments=inactive&text_search=taylor HTTP/1.1" 200 3466
Traceback (most recent call last):
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/wsgiref/handlers.py", line 85, in run
    self.result = application(self.environ, self.start_response)
  File "/Users/dsjen/Envs/edx-analytics-dashboard/lib/python2.7/site-packages/django/contrib/staticfiles/handlers.py", line 63, in __call__
    return self.application(environ, start_response)
  File "/Users/dsjen/Envs/edx-analytics-dashboard/lib/python2.7/site-packages/django/core/handlers/wsgi.py", line 185, in __call__
    start_response(force_str(status), response_headers)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/wsgiref/handlers.py", line 181, in start_response
    assert not is_hop_by_hop(name),"Hop-by-hop headers not allowed"
AssertionError: Hop-by-hop headers not allowed
[23/Aug/2016 18:08:54] "GET /api/learner_analytics/v0/learners.csv?course_id=[removed/course/id]&enrollment_mode=honor&ignore_segments=inactive&text_search=taylor HTTP/1.1" 500 59
```

Have you seen this before?  I'm still early in the process of getting this up, but wanted to check.

dsjen commented:
@pomegranited -- I noticed in the testing instructions that one of the requested fields is "first_enrollment_date".  I don't see that field in https://github.com/edx/edx-analytics-data-api/pull/129.  Is that missing from the PR?

dsjen commented:
@pomegranited -- I was able to test this out and download a CSV with cohort, enrollment, segment, and search filters applied.  🙌  Hooray! 🙌 

I noticed that the CSV downloaded only has the first 25 learners ever though.  Is that expected?  I thought all learners were to be downloaded.

thallada commented on analytics_dashboard/static/apps/learners/common/views/download-data.js:
```diff
@@ -0,0 +1,81 @@
+
+            // Assemble a sorted list of query parameters from the active filters
+            _.extend(queryParams, this.download_params,
+                                  this.options.collection.getActiveFilterFields(true));
```
The page parameter is missing from these queryParams, which means no matter what page of the roster I'm on, it will return only the first page in the CSV. In addition to that, `page_size`, `order_by`, and `sort_order` should be passed through as well. Those are respectively stored in the collection state as `pageSize`, `sortKey`, and `order`. The translations for those are specified by the collection's queryParams variable. See https://github.com/backbone-paginator/backbone.paginator#adapting-to-a-server-api.

There's a broader question here about whether we want to download CSVs one page at a time. In all of our other reports, the CSV option downloads the entire set of data unpaginated. @dsjen @stroilova

pomegranited replied:
@thallada @dsjen @stroilova 

[92f7abe](https://github.com/open-craft/edx-analytics-dashboard/commit/92f7abee296d3111089fb5a8e7f5fce22a14d852) adds `collection.getQueryString()` to the CSV download URL, which includes `page`, `order_by` and `sort_order`, as set by the search form.

But I'm unsure what to do about `page_size`.

The Analytics API is generating this downloadable CSV file, and by default, it has `DEFAULT_PAGE_SIZE=25`, and `MAX_PAGE_SIZE: 100`.  These can be [configured higher](https://github.com/open-craft/configuration/blob/899ec486b6aaedeebc02ac64fc7081624e9b7d26/playbooks/roles/analytics_api/defaults/main.yml#L67-L68), and for clients with a small number of learners, this is sufficient.

However, what to do for large courses?  The Analytics API's CSV responses follow the [github API convention](https://developer.github.com/guides/traversing-with-pagination/) for including [pagination data in the response header](https://github.com/edx/edx-analytics-data-api/pull/128), e.g.  

```
Link:<http://127.0.0.1:8100/api/v0/learners/?course_id=course-v1%3AOpenCraft%2BAnalytics101x%2Bnow&page=2>; rel="next"
```

So in theory, the Insights `Download CSV` button could have prev/next download buttons, but that starts making the UI complex.

Any ideas?

pomegranited replied:
@thallada @dsjen @stroilova 

I've dealt with the page_size problem by using the collection's totalRecords count as the page_size parameter to the Analytics API.  Note that this will only work if the API has a max_page_size >= page_size, but at least it should get us all the records the API will give.

See [a421e4c](https://github.com/edx/edx-analytics-dashboard/pull/533/commits/a421e4cf859f0288878542a2fb52e10b86153774).

thallada replied:
Great! Because this feature is under a waffle switch, it should be good to merge. We'll have to do some load testing on a course with a large number of learners and the max_page_size turned off to see whether we can enable this for all edx.org courses.

pomegranited commented:
@dsjen The `first_enrollment_date` field was a typo.  It's just `enrollment_date`, and it was already present in the engagement data, so I didn't have to add it.

I've updated this PR description to fix the field name.

pomegranited commented:
@dsjen @thallada Warning: I think I need to rebase from master to get the automated tests to run again..  So I'll be editing all those comments I just made to point to the correct changesets.

So if you're looking at this PR right now, you might lose your place.  My apologies!

dsjen commented:
I ran into the same build error that this PR is currently having.  It seems like data can't be generated for the data api, and I'm looking into it right now.  We made a change to the data api recently, and I wonder if something went awry.

dsjen commented:
@pomegranited -- thanks for all the updates!  It looks great to me, and I'm very excited about this functionality.  I looked at the build thinking there was a flaky test, but it looks like a JS test is failing.

I believe there is one outstanding question concerning the download CSV button vs. link that is awaiting input.

Other than that, LGTM! 👍

pomegranited commented:
Squashed 1905a08..e3e96b4 into 210ddc7.

clrux commented on analytics_dashboard/static/apps/learners/common/templates/download-data.underscore:
```diff
@@ -0,0 +1,10 @@
+<% if (hasDownloadData) { %>
+    <div class="section-action">
+        <a class="btn btn-default action-download-data" role="button"
+           href="#" data-role="learner-data-csv" data-track-type="click"
```
Is there any way to use an actual `href` value here, perhaps to the filename? This will trigger a11y test failures.

clrux replied:
Also, providing a value will ensure this works even if JavaScript isn't working or is unavailable.

pomegranited replied:
@clrux Excellent point! 3144150 changes the template to take the initial `downloadUrl` when the template is generated, instead of setting it with Javascript.

If Javascript is disabled, then subsequent updates to the form won't change the Download CSV URL, but at least it will have a valid URL.

clrux commented:
Thanks! :+1:

pomegranited commented:
I've addressed @clrux's comments, and fixed the affected tests.

Squashed 3f86edd..3144150 into 3914876.

@thallada and @dsjen was there anything else you need me to fix before this can be merged?

dsjen commented:
@pomegranited -- thank you for the diligence in pushing this forward and for squashing the commits.  This looks great to me!  @thallada, how's it look from your viewpoint?

thallada commented on analytics_dashboard/static/download-form/diff:
```diff
@@ -0,0 +1,182 @@
+diff --git a/Makefile b/Makefile
```
Was this file accidentally included?

pomegranited replied:
YES!  My apologies.  Removed with 4144303.

thallada commented:
Just had one minor comment. Everything else looks good to me :+1:. Thanks for addressing all of our comments!

pomegranited commented:
Thanks for catching that accidental add, @thallada !

All good now?

dsjen commented:
@pomegranited -- thanks for all the work on this.  We really appreciate you extending functionality on Insights!
